### PR TITLE
added DAO and implementation for CinemaHall and MovieSession

### PR DESCRIPTION
**Pull request by mikhail771**

### CODE REVIEW

File: src/main/java/com/dev/cinema/dao/impl/CinemaHallDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.save(cinemaHall);
+            transaction.commit();
+            return cinemaHall;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't insert Cinema Hall entity", e);
```
Comment (dariiabutorina) [inline src/main/java/com/dev/cinema/dao/impl/CinemaHallDaoImpl.java]:
```suggestion
            throw new DataProcessingException("Can't insert Cinema Hall entity " + cinemaHall, e);
```

Comment (dariiabutorina) [reply src/main/java/com/dev/cinema/dao/impl/CinemaHallDaoImpl.java]: Let's make messages more informative

File: src/main/java/com/dev/cinema/dao/impl/CinemaHallDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    @Override
+    public List<CinemaHall> getAll() {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<CinemaHall> getAllHallsQuery =
+                    session.createQuery("from CinemaHall", CinemaHall.class);
+            return getAllHallsQuery.getResultList();
+        } catch (Exception e) {
+            throw new DataProcessingException("Failed to load cinema halls", e);
+        }
+    }
```
Comment (dariiabutorina) [inline src/main/java/com/dev/cinema/dao/impl/CinemaHallDaoImpl.java]:
```suggestion
        }
    }
```

Comment (dariiabutorina) [reply src/main/java/com/dev/cinema/dao/impl/CinemaHallDaoImpl.java]: No need in handling exceptions in the getAll() methods

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                    + "join fetch ms.movie m "
+                    + "where m.id = :movieId "
+                    + "and ms.sessionTime between :from AND :to", MovieSession.class);
+            query.setParameter("movieId", movieId);
+            query.setParameter("from", date.atTime(LocalTime.MIN));
+            query.setParameter("to", date.atTime(LocalTime.MAX));
+            return query.getResultList();
+        } catch (Exception e) {
+            throw new DataProcessingException("Failed to find available sessions", e);
+        }
```
Comment (dariiabutorina) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: Same

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import java.time.LocalTime;
+import java.util.List;
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+import org.hibernate.query.Query;
+
+@Dao
+public class MovieSessionDaoImpl implements MovieSessionDao {
+
+    @Override
```
Comment (dariiabutorina) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]:
```suggestion
public class MovieSessionDaoImpl implements MovieSessionDao {
    @Override
```

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.save(movieSession);
+            transaction.commit();
+            return movieSession;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't insert Movie Session entity", e);
```
Comment (dariiabutorina) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: Same

File: src/main/java/com/dev/cinema/model/MovieSession.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import javax.persistence.OneToOne;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "movie_session")
+public class MovieSession {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
```
Comment (dariiabutorina) [inline src/main/java/com/dev/cinema/model/MovieSession.java]:
```suggestion
    @ManyToOne
```

Comment (dariiabutorina) [reply src/main/java/com/dev/cinema/model/MovieSession.java]: One movie may remain shown several times - have several movie sessions

**Review by dariiabutorina — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/cinema/dao/impl/CinemaHallDaoImpl.java (hunk 1 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/dao/impl/CinemaHallDaoImpl.java]: done

File: src/main/java/com/dev/cinema/dao/impl/CinemaHallDaoImpl.java (hunk 2 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/dao/impl/CinemaHallDaoImpl.java]: done

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java (hunk 3 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: done

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java (hunk 4 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: done

File: src/main/java/com/dev/cinema/model/MovieSession.java (hunk 6 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/model/MovieSession.java]: done

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java (hunk 5 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: done

**Comment by RomanDubovskyi** [comment]
@DariiaPikul good comments, nice 👍

File: src/main/java/com/dev/cinema/model/MovieSession.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+@Entity
+@Table(name = "movie_session")
+public class MovieSession {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
+    private Movie movie;
+    @ManyToOne
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/model/MovieSession.java:19]: this is for your info, but keep in mind that toOne relations have fetchType eager by default, which is not always quite good

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**